sdk/program/src/message/original.rs

```diff
@@ -163,28 +164,7 @@ fn get_program_ids(instructions: &[Instruction]) -> Vec<Pubkey> {
         .collect()
 }
 
-pub const MESSAGE_HEADER_LENGTH: usize = 3;
-
-#[frozen_abi(digest = "BVC5RhetsNpheGipt5rUrkR6RDDUHtD5sCLK1UjymL4S")]
-#[derive(Serialize, Deserialize, Default, Debug, PartialEq, Eq, Clone, AbiExample)]
-#[serde(rename_all = "camelCase")]
-pub struct MessageHeader {
-    /// The number of signatures required for this message to be considered valid. The
-    /// signatures must match the first `num_required_signatures` of `account_keys`.
-    /// NOTE: Serialization-related changes must be paired with the direct read at sigverify.
-    pub num_required_signatures: u8,
-
-    /// The last num_readonly_signed_accounts of the signed keys are read-only accounts. Programs
-    /// may process multiple transactions that load read-only accounts within a single PoH entry,
-    /// but are not permitted to credit or debit lamports or modify account data. Transactions
-    /// targeting the same read-write account are evaluated sequentially.
-    pub num_readonly_signed_accounts: u8,
-
-    /// The last num_readonly_unsigned_accounts of the unsigned keys are read-only accounts.
-    pub num_readonly_unsigned_accounts: u8,
-}
-
-#[frozen_abi(digest = "BPBJZxpRQ4JS7LGJtsgoyctg4BXyBbbY4uc7FjowtxLV")]
+#[frozen_abi(digest = "8Rf1mAQbCgnTLkS4CUK9Tx2i7puRp3aq1qw6AGZCrj1L")]
```

Comment:
This changed because the ABI macro uses the fully qualified struct name in its hash (`solana_program::message::Message` vs `solana_program::message::original::Message`)

Reply:
I have a change here that would prevent this: https://github.com/solana-labs/solana/pull/18734

Reply:
@jstarry it's no surprise you created #18734 because of lack of design philosophy of frozen_abi... I constantly fail to empty my work queue to document it properly.... xD

well, frozen_abi is designed to catch even structurally-equal rename like this, forcing devs to check their sanity.
so, I'd rather keep this noisy (i know...) behavior.

I'm aware it's bothersome and no fun to be suspected for possibly breaking abi for well-sanctioned just _rename_s. but frozen_abi is tasked for this noisy policing in that way.

Sympathizing human, frozen_abi wants to minimize the maintenance burden of each and every safe abi changes. So, frozen_abi is designed to be used sparingly across the code base, ideally each for the topmost type of some hierarchy with abi concerns.

To that end, frozen_abi tries _super_ hard to structurally digest type information even piercing `enum` variants separately into each arm (and recursively). (I guess you felt the die-hardness of the frozen_code..) This hassle is all for to make fewer and more wider frozen_abi possible for the purpose.

in other words, the versioned transaction should comprise the single coherent frozen_abi hash now as a whole from frozen_abi's semantic viewpoint. For example, adding yet another newer next-gen tx format surely would break this abi, so versioning must be part of it. Likewise, even single tiny change deepest in the type hierarchy should be caught by the sole and mighty frozen_abi residing at the top versioning `enum`. That's the motto of frozen_abi.

So to recap, `MessageHeader` is anti-pattern (sorry this was confusing to be begin with) and `Protocol` is good-pattern according to the philosophy. And I left comments on this pr according to this thinking.